lib/package.js

```diff
@@ -760,24 +761,40 @@ exports.createMain = function(pkg, pjson, downloadDir) {
     }
 
     if (main) {
-      if (main.startsWith('./'))
-        main = main.substr(2);
-      if (main.endsWith('.js'))
-        main = main.substr(0, main.length - 3);
+      if (main.startsWith('./')) {
+        main = main.substr(2);    
+      }
+    }
+    else {
+      main = 'index' + mainFileExtension;
     }
 
-    main = main || 'index';
+    return new Promise(function(resolve) {
+      // Check to see if main points to an actual file or not
+      // i.e. does it include extension.
+      mainPath = path.resolve(downloadDir, main);
+      fs.exists(mainPath, resolve);
+    });
+
+  }).then(function(exists) {
+    if (exists) {
+      var match = /(\.[^.]+)$/.exec(main);
+      if (match) {
+        mainFileExtension = match[1];
+        main = main.slice(0, - mainFileExtension.length);
```

Comment:
I'm not sure I understand this - why do you need to remove the extension at all?